### ISSUE TITLE
Problem with identifying two components

### ISSUE THREAD

**Issue by mitibabaei** [gausspy/gausspy#17]
Hi, I am trying to fit my spectrum with multiple components. I used the exact sample script with multiple numbers of components. I set alpha1 = 1.0 and snr_thresh = 0.01 . Unfortunately, it doesn't work and I could not find any other parameters that help. I attached two sample spectrums and their plots here. Could you please check them and help me how can I fix this problem?
![Screenshot from 2020-10-11 23-42-51](https://user-images.githubusercontent.com/39658321/95691110-f2f9b200-0c1c-11eb-9e00-7ccdbf8ad255.png)
![Screenshot from 2020-10-11 23-41-54](https://user-images.githubusercontent.com/39658321/95691116-0442be80-0c1d-11eb-864d-3a7778355087.png)
[al_fil4_15.txt](https://github.com/gausspy/gausspy/files/5362212/al_fil4_15.txt)
[ap_fil4_30.txt](https://github.com/gausspy/gausspy/files/5362213/ap_fil4_30.txt)

The attached code is also the script I have used.

```
#######################################################
#define gaussion function #############################
def gaussion(amp, fwhm, mean):
	return lambda x: amp * np.exp(-4 * np.log(2) * (x-mean)**2 / fwhm**2)

def unravel(list):
	return np.array([i for array in list for i in array])

#######################################################
#read txt file#########################################
for i in range(1, 2):
	spec=np.loadtxt('ap_fil4_'+str(i)+'.txt')
	velocity = spec[:,0]
	intensity = spec[:,1]
	errors = np.ones(len(intensity)) * 0.001

	#######################################################
	#calc num of components ###############################
	peak_positive, _ = find_peaks(intensity, height = 1, threshold = None, distance = 1)
	for k in peak_positive:
		print velocity[k]

	#######################################################
	#create pickle file####################################
	FILENAME = 'multiple_gaussians.pickle'

	#Number of Gaussion functions per spectrum
	ncomp = len(peak_positive)
	print('number of components =', ncomp)

	#Enter results into AGD dataset
	data={}
	data['data_list'] = data.get('data_list', []) + [intensity]
	data['x_values'] = data.get('x_values', []) + [velocity]
	data['errors'] = data.get('errors', []) + [errors]

	pickle.dump(data, open(FILENAME, 'w'))

	#######################################################
	#Run Gausspy###########################################

	#specify necessary parameters
	alpha1 = 1.0
	snr_thresh = 0.01
	FILENAME_DATA = 'multiple_gaussians.pickle'
	FILENAME_DATA_DECOMP = 'multiple_gaussians_decomposed.pickle'

	#load GaussPy
	g = gp.GaussianDecomposer()

	#Setting AGD parameters
	g.set('phase', 'one')
	g.set('snr_thresh', [snr_thresh, snr_thresh])
	g.set('alpha1', alpha1)

	#Run GaussPy
	data_decomp = g.batch_decomposition(FILENAME_DATA)

	#Save decomposition information 
	pickle.dump(data_decomp, open(FILENAME_DATA_DECOMP, 'w'))

	#######################################################
	#Plot the result#######################################
	data_decomp = pickle.load(open(FILENAME_DATA_DECOMP))

	means_fit = unravel(data_decomp['means_fit'])
	amps_fit = unravel(data_decomp['amplitudes_fit'])
	fwhms_fit = unravel(data_decomp['fwhms_fit'])

	fig = plt.figure()
	ax = fig.add_subplot(111)

	model = np.zeros(len(velocity))

	for j in range(len(means_fit)):
		component = gaussion(amps_fit[j], fwhms_fit[j], means_fit[j])(velocity)
		model += component

	ax.plot(velocity, intensity, label='Data', color='gray', linewidth=1.5)
	ax.plot(velocity, model, label = r'$\log\alpha=1.$', color='turquoise', linewidth=2.)
	ax.set_xlabel('Velocity')
	ax.set_ylabel('Amplitude')

	np.savetxt('ap_fil4_'+str(i)+'_fit'+'.txt', np.c_[velocity, model])
	plt.savefig('ap_fil4_'+str(i)+'.png' )

#plt.show()
```

**Comment by cmurray-astro** [comment]
Thanks for your question. Have you tried generating a synthetic dataset (or using a real dataset with known decomposition) to train your value of alpha? Also, when you say that you could not find other parameters that help, do you mean that you tried other values of alpha and did not see changes in the decomposition? Another idea is to make sure your estimate of the noise is accurate (which is used in the signal to noise (S/N) threshold).

**Comment by mitibabaei** [comment]
Thanks for your reply. 
As you suggested, I defined two alpha values and train my data set before using Gausspy but still does not work. This is the code I have used: 


`# Create training dataset with Gaussian profiles

import numpy as np
import pickle
import gausspy.gp as gp
#plotting packages
import matplotlib
import matplotlib.pyplot as plt
import pyfits
from astropy.io import fits
import astropy.io.ascii as ascii 
import os
import numpy as np
from astropy import units as u
import pandas as pd
import matplotlib.gridspec as gridspec
import aplpy
from scipy.signal import find_peaks
#read txt file#########################################
spec = np.loadtxt('/scratch/atefeh/projects/NGC6334/source-V/Combine/spw7_h13co+/pv_plots/spectrum_txt/apex/fil4/ap_fil4_30.txt')
velocity = spec[:,0]
intensity = spec[:,1]
errors = np.ones(len(intensity)) * 0.001

#######################################################
#calc num of components ###############################
peak_positive, _ = find_peaks(intensity, height = 1, threshold = None, distance = 1)
for i in peak_positive:
	print velocity[i]

#######################################################
#define gaussion function #############################
def gaussion(amp, fwhm, mean):
	return lambda x: amp * np.exp(-4 * np.log(2) * (x-mean)**2 / fwhm**2)

def unravel(list):
	return np.array([i for array in list for i in array])

#######################################################
#create pickle file####################################
FILENAME = 'multiple_gaussians.pickle'

#Number of Gaussion functions per spectrum
NCOMPS = len(peak_positive)
print('NCOMPS:', NCOMPS)

#Enter results into AGD dataset
data={}
data['data_list'] = data.get('data_list', []) + [intensity]
data['x_values'] = data.get('x_values', []) + [velocity]
data['errors'] = data.get('errors', []) + [errors]
TRAINING_SET = True
pickle.dump(data, open(FILENAME, 'w'))

#######################################################
#######################################################
# Select the optimal value of alpha by training the AGD algorithm
# Set necessary parameters
FILENAME_TRAIN = 'training_data.pickle'
snr_thresh = 5.
alpha1_initial = 0.28
alpha2_initial = 2.42

g = gp.GaussianDecomposer()

# Next, load the training dataset for analysis:
g.load_training_data(FILENAME_TRAIN)

# Set GaussPy parameters
g.set('phase', 'two')
g.set('SNR_thresh', [snr_thresh, snr_thresh])

# Train AGD starting with initial guess for alpha
g.train(alpha1_initial = alpha1_initial, alpha2_initial = alpha2_initial)
#######################################################
# Decompose multiple Gaussian dataset using AGD with TRAINED alpha

# Specify necessary parameters
alpha1 = alpha1_initial
alpha2 = alpha2_initial
snr_thresh = 1.

FILENAME_DATA = 'multiple_gaussians.pickle'
FILENAME_DATA_DECOMP = 'multiple_gaussians_trained_decomposed.pickle'

# Load GaussPy
g = gp.GaussianDecomposer()

# Setting AGD parameters
g.set('phase', 'one')
g.set('SNR_thresh', [snr_thresh, snr_thresh])
g.set('alpha1', alpha1)
g.set('alpha2', alpha2)
# Run GaussPy
data_decomp = g.batch_decomposition(FILENAME_DATA)

# Save decomposition information
pickle.dump(data_decomp, open(FILENAME_DATA_DECOMP, 'w'))

#######################################################
#Plot the result#######################################


data_decomp = pickle.load(open(FILENAME_DATA_DECOMP))


means_fit = unravel(data_decomp['means_fit'])
amps_fit = unravel(data_decomp['amplitudes_fit'])
fwhms_fit = unravel(data_decomp['fwhms_fit'])

fig = plt.figure()
ax = fig.add_subplot(111)

model = np.zeros(len(velocity))

for j in range(len(means_fit)):
	component = gaussion(amps_fit[j], fwhms_fit[j], means_fit[j])(velocity)
	model += component

ax.plot(velocity, intensity, label='Data', color='gray', linewidth=1.5)
ax.plot(velocity, model, label = r'$\log\alpha=1.$', color='turquoise', linewidth=2.)
ax.set_xlabel('Velocity')
ax.set_ylabel('Amplitude')

#print means_fit
#print amps_fit
#print fwhms_fit
#np.savetxt('../docs_txt/'+str(30)+'_fit'+'.txt', np.c_[velocity, model])
#plt.savefig('/scratch/atefeh/projects/NGC6334/source-V/gausspy/plots/testapex30.png', bbox_inches='tight', dpi=300 )

plt.show()`

**Comment by mitibabaei** [comment]
Thanks for your reply.
As you suggested, I defined two alpha values and train my data set before using Gausspy but still does not work. This is the result I get from training:
```
0.27720065452411424 0.9 0.0 -0.0020792394857790607
iter 31: F1=76.7%, alpha=[0.28, 2.41], p=[-0.00, -0.00]   (0.00,0.00 < 0.03 for 11 iters [10 required])
Stable convergence acheived at iteration:  31
ilist =  [0]
   ---->   0
/usr/lib64/python2.7/site-packages/gausspy/AGD_decomposer.py:245: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  amps_new = lstsq(FF_matrix,amps)[0]
SUCCESS
100 finished.%
```


I also attached the part of modified scripts here, maybe I've used alpha parameters in a wrong format. 

```
#specify necessary parameters
alpha1 = 0.28
alpha2 = 2.41
snr_thresh = 0.01
FILENAME_DATA = 'multiple_gaussians.pickle'
FILENAME_DATA_DECOMP = 'multiple_gaussians_decomposed.pickle'

#load GaussPy
g = gp.GaussianDecomposer()

#Setting AGD parameters
g.set('phase', 'one')
g.set('snr_thresh', [snr_thresh, snr_thresh])
g.set('alpha1', alpha1)
g.set('alpha2', alpha2)
#Run GaussPy
data_decomp = g.batch_decomposition(FILENAME_DATA)

#Save decomposition information 
pickle.dump(data_decomp, open(FILENAME_DATA_DECOMP, 'w'))

```


.

**Comment by cmurray-astro** [comment]
Can you describe what you mean by "doesn't work"? Do you mean that you expect two components but you still get only one? How did you build your training data set? 

Beyond that, I think one issue is with the snr_thresh: this is defined as the ratio of the signal to the noise, so setting it to 0.01 as you have requires the signal to be 100x less than the noise. For another suggestion, in the GaussPy paper (Linder et al. 2015: https://ui.adsabs.harvard.edu/abs/2015AJ....149..138L/abstract) we describe re-sampling our spectra to smaller per channel resolution (e.g., sampling from 0.4 km/s per channel to 0.1 km/s per channel velocity resolution) to avoid aliasing narrow components. You might try this as well.

**Comment by mitibabaei** [comment]
Actually, the training algorithm yielded two quantities for alpha but Gausspy code only fit one single component. the attached plot is the result of setting two alpha and as you can see, pretty identical to the one I sent earlier. 
![Screenshot from 2020-10-15 12-22-31](https://user-images.githubusercontent.com/39658321/96222655-f7133000-0f8c-11eb-9aa2-952823353c25.png)

**Comment by cmurray-astro** [comment]
I recommend reading through Lindner et al. 2015 (https://ui.adsabs.harvard.edu/abs/2015AJ....149..138L/abstract) for more background on how the alpha parameters are chosen (and to decide between "one phase" and "two phase" for your application). How you build the training set matters -- it must be representative of your data, and so please check that your training data is accurate.
